### DISCUSSION TITLE
SQL configuration

### DISCUSSION THREAD

**Discussion by timofeyp** [strongswan/strongswan#1263] [Q&A]
I've installed strongswan 5.9.7 with sql plugin and configured it in `strongswan.conf` charon plugins section:

`charon {
	load = charon-systemd test-vectors ldap pkcs11 tpm aesni aes rc2 sha2 sha1 md5 mgf1 rdrand random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl gcrypt af-alg fips-prf gmp curve25519 agent chapoly xcbc cmac hmac ctr ccm gcm ntru drbg curl attr kernel-netlink resolve socket-default connmark forecast farp vici updown eap-identity eap-aka eap-md5 eap-gtc eap-mschapv2 eap-dynamic eap-radius eap-tls eap-ttls eap-peap eap-tnc xauth-generic xauth-eap xauth-pam tnc-tnccs dhcp lookip error-notify certexpire led addrblock unity counters sql mysql
	plugins {
		include strongswan.d/charon/*.conf
		sql {
  			database = mysql://root:PASSWORD@localhost/strongswan
  			loglevel = 1
		}
                attr-sql {
                    database = mysql://root:*PASSWORD@localhost/strongswan
                }
	}
}

include strongswan.d/*.conf`

SQL schema is [https://github.com/strongswan/strongswan/blob/master/src/pool/sqlite.sql](https://github.com/strongswan/strongswan/blob/master/src/pool/sqlite.sql)

It seems sql connection works, because **logs** table fills up records depends of loglevel param.
But if im trying to add some certificates to **certificates** table, I cant see it by command `swanctl --list-certs`. Only disk stored certs shown. Also I've tried to import full examples sql configurations from [https://strongswan.org/testresults.html](https://strongswan.org/testresults.html), and no any settings of db seen after command `swanctl --load-all`, but only settings of my  `/etc/swanctl/swanctl.conf`. 
How to switch strongswan to load configuration from database instead of `/etc/swanctl/swanctl.conf`

**Comment by Thermi** [comment]
swanctl only lists things loaded through the VICI API which it uses for configuration management. swanctl doesn't know about the SQL DB.
EDIT: except connection related settings.

**Reply by timofeyp** [reply]
But certificates listed by swanctl, after I restart strongswan service.


```
root@vm64326 ~ # swanctl --list-certs

List of X.509 End Entity Certificates

  subject:  "CN=45.11.183.161"
  issuer:   "CN=VPN root CA"
  validity:  not before Jul 24 16:41:51 2022, ok
             not after  Jul 23 16:41:51 2027, ok (expires in 1787 days)
  serial:    46:e3:fc:c0:00:49:6c:15
  altNames:  45.11.183.161, 45.11.183.161
  flags:     serverAuth ikeIntermediate 
  authkeyId: f3:02:c8:c2:4f:3a:8b:ff:ca:3e:4c:5f:db:1a:fb:95:25:da:c3:b0
  subjkeyId: 8c:9a:0b:d6:1c:dc:55:16:d1:92:85:76:f0:20:9c:13:72:d3:f4:46
  pubkey:    RSA 4096 bits, has private key
  keyid:     98:d1:f6:47:dc:db:16:38:36:e6:72:7b:bc:d9:c1:e7:02:5d:21:18
  subjkey:   8c:9a:0b:d6:1c:dc:55:16:d1:92:85:76:f0:20:9c:13:72:d3:f4:46

```

Then I do restart:
```
root@vm64326 ~ # systemctl restart strongswan
root@vm64326 ~ # swanctl --list-certs

List of X.509 End Entity Certificates

  subject:  "CN=45.11.183.161"
  issuer:   "CN=VPN root CA"
  validity:  not before Jul 24 16:41:51 2022, ok
             not after  Jul 23 16:41:51 2027, ok (expires in 1787 days)
  serial:    46:e3:fc:c0:00:49:6c:15
  altNames:  45.11.183.161, 45.11.183.161
  flags:     serverAuth ikeIntermediate 
  authkeyId: f3:02:c8:c2:4f:3a:8b:ff:ca:3e:4c:5f:db:1a:fb:95:25:da:c3:b0
  subjkeyId: 8c:9a:0b:d6:1c:dc:55:16:d1:92:85:76:f0:20:9c:13:72:d3:f4:46
  pubkey:    RSA 4096 bits, has private key
  keyid:     98:d1:f6:47:dc:db:16:38:36:e6:72:7b:bc:d9:c1:e7:02:5d:21:18
  subjkey:   8c:9a:0b:d6:1c:dc:55:16:d1:92:85:76:f0:20:9c:13:72:d3:f4:46

  subject:  "C=CH, O=strongSwan Project, CN=moon.strongswan.org"
  issuer:   "C=CH, O=strongSwan Project, CN=strongSwan Root CA"
  validity:  not before Oct 20 07:50:47 2021, ok
             not after  Oct 20 07:50:47 2029, ok (expires in 2607 days)
  serial:    03
  altNames:  moon.strongswan.org
  flags:     
  CRL URIs:  http://crl.strongswan.org/strongswan.crl
  authkeyId: 9c:b0:d5:d5:44:ce:48:6d:72:b1:b6:aa:49:fb:4a:50:81:87:ae:a6
  subjkeyId: b7:fe:93:f5:0a:ff:d1:0b:cc:14:74:8d:43:48:b3:0b:e5:4c:6f:73
  pubkey:    RSA 3072 bits
  keyid:     79:41:59:0c:e5:a0:fa:03:97:df:a3:17:da:98:21:1d:fd:10:cd:8f
  subjkey:   b7:fe:93:f5:0a:ff:d1:0b:cc:14:74:8d:43:48:b3:0b:e5:4c:6f:73

List of X.509 CA Certificates

  subject:  "C=CH, O=strongSwan Project, CN=strongSwan Root CA"
  issuer:   "C=CH, O=strongSwan Project, CN=strongSwan Root CA"
  validity:  not before Oct 20 07:50:47 2021, ok
             not after  Oct 21 07:50:47 2031, ok (expires in 3338 days)
  serial:    33:0d:91:9b:d9:bb:a3:f1
  flags:     CA CRLSign self-signed 
  pathlen:   1
  subjkeyId: 9c:b0:d5:d5:44:ce:48:6d:72:b1:b6:aa:49:fb:4a:50:81:87:ae:a6
  pubkey:    RSA 3072 bits
  keyid:     be:bc:f0:fd:0c:82:e4:82:01:69:b2:50:25:34:0d:80:0b:fd:b3:17
  subjkey:   9c:b0:d5:d5:44:ce:48:6d:72:b1:b6:aa:49:fb:4a:50:81:87:ae:a6
```

Why certificates from DB listed by swanctl after restarting strongswan service ? Can strongswan apply configuration from db without restarting ?

**Reply by Thermi** [reply]
Weird, seems I did not dig deep enough in the code then.
I did not check, but maybe `swanctl --reload-settings` works for the SQL plugin

**Reply by tobiasbrunner** [reply]
No config reload or restart is necessary to us new certificates or configs from the DB. But changing the connection URIs requires a restart, `--reload-settings` is currently not enough.

**Comment by Thermi** [comment]
Shouldn't they be immediately visible then though?
From what I gathered from the post wnd hi comments, they are not.

Am 1. September 2022 07:03:43 UTC schrieb Tobias Brunner ***@***.***>:
>No config reload or restart is necessary to us new certificates or configs from the DB. But changing the connection URIs requires a restart, `--reload-settings` is currently not enough.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/strongswan/strongswan/discussions/1263#discussioncomment-3524032
>You are receiving this because you commented.
>
>Message ID: ***@***.***>
-- 
Sent from mobile. Please excuse any typos or brevity
